### PR TITLE
feat: Widget grouping Extended Features

### LINKED ISSUE #5995: [Feature] Allow the multi-select polygon to be drawn from outside the canvas
Today the polygon can only be drawn inside the canvas, but it's difficult to select widgets that are edge to edge. There's no empty space to draw the polygon.

### LINKED ISSUE #6102: [Feature] cmd+a on a selected container should select all children of the container-like widget

### PR DESCRIPTION
**Pull request by marks0351**
## Description

- [x] Draw to select from outside of the editor.
- [x] Cmd + A to select child widgets inside containers, forms, list, tabs, and modals.

Fixes #5995
Fixes #6102

> if no issue exists, please create an issue and ask the maintainers about this first

## Type of change

> Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## How Has This Been Tested?

> Please describe the tests that you ran to verify your changes. Provide instructions, so we can reproduce.
> Please also list any relevant details for your test configuration.

- Test A
- Test B

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feature/widget-group-extend 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.14 **(0.6)** | 37.27 **(0.86)** | 33.8 **(0.92)** | 55.72 **(0.57)**
 :green_circle: | app/client/src/actions/batchActions.ts | 100 **(28.57)** | 100 **(0)** | 100 **(100)** | 100 **(0)**
 :green_circle: | app/client/src/actions/metaActions.ts | 81.82 **(9.09)** | 100 **(0)** | 33.33 **(33.33)** | 75 **(12.5)**
 :green_circle: | app/client/src/actions/widgetSelectionActions.ts | 100 **(0)** | 100 **(50)** | 100 **(0)** | 100 **(0)**
 :sparkles: :new: | **app/client/src/components/designSystems/appsmith/ChartComponent.tsx** | **47.1** | **21.43** | **46.67** | **45.14**
 :green_circle: | app/client/src/components/designSystems/appsmith/ContainerComponent.tsx | 93.1 **(0)** | 65 **(2.5)** | 100 **(0)** | 93.1 **(0)**
 :green_circle: | app/client/src/components/designSystems/blueprint/ButtonComponent.tsx | 57.26 **(9.4)** | 50.5 **(10.9)** | 55.56 **(0)** | 57.94 **(6.54)**
 :green_circle: | app/client/src/components/designSystems/blueprint/TextComponent.tsx | 100 **(56.25)** | 73.33 **(63.33)** | 100 **(90.91)** | 100 **(58.06)**
 :red_circle: | app/client/src/constants/messages.ts | 70.12 **(-0.12)** | 100 **(0)** | 9.82 **(-0.36)** | 77.01 **(-0.26)**
 :green_circle: | app/client/src/pages/Editor/WidgetsEditor.tsx | 91.67 **(0.9)** | 71.88 **(0)** | 100 **(0)** | 91.3 **(0.98)**
 :green_circle: | app/client/src/pages/common/CanvasSelectionArena.tsx | 91.35 **(1.64)** | 71.82 **(4.85)** | 89.66 **(7.84)** | 91.06 **(1.59)**
 :green_circle: | app/client/src/reducers/entityReducers/metaReducer.ts | 42.86 **(10.72)** | 0 **(0)** | 28.57 **(14.28)** | 42.86 **(10.72)**
 :green_circle: | app/client/src/sagas/BatchSagas.tsx | 87.1 **(61.29)** | 91.67 **(83.34)** | 100 **(80)** | 88 **(56)**
 :green_circle: | app/client/src/sagas/EvaluationsSaga.ts | 21.1 **(0)** | 3.13 **(0)** | 13.33 **(6.66)** | 24.47 **(0)**
 :green_circle: | app/client/src/sagas/SelectionCanvasSagas.ts | 100 **(6.82)** | 72.22 **(15.97)** | 100 **(37.5)** | 100 **(7.69)**
 :green_circle: | app/client/src/sagas/WidgetOperationSagas.tsx | 56.32 **(0.15)** | 38.83 **(0.32)** | 58.14 **(0)** | 56.22 **(0)**
 :green_circle: | app/client/src/sagas/WidgetOperationUtils.ts | 66.67 **(11.39)** | 44.44 **(10.1)** | 76.92 **(7.69)** | 65.45 **(11.81)**
 :red_circle: | app/client/src/sagas/WidgetSelectionSagas.ts | 91.1 **(-5.37)** | 87.1 **(-4.9)** | 100 **(0)** | 90.4 **(-5.65)**
 :green_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 74.03 **(1.85)** | 66.2 **(1.8)** | 69.23 **(2.2)** | 73.1 **(1.95)**
 :green_circle: | app/client/src/utils/WidgetRegistry.tsx | 80.18 **(4.5)** | 100 **(0)** | 40.54 **(13.51)** | 80.18 **(4.5)**
 :green_circle: | app/client/src/utils/helpers.tsx | 64.74 **(1.73)** | 35.16 **(1.09)** | 41.38 **(6.9)** | 61.59 **(2.17)**
 :green_circle: | app/client/src/utils/hooks/useBlocksToBeDraggedOnCanvas.ts | 95.73 **(0)** | 85.71 **(1.42)** | 86.36 **(0)** | 95.5 **(0)**
 :green_circle: | app/client/src/widgets/BaseWidget.tsx | 76.23 **(0)** | 44.19 **(6.98)** | 71.88 **(0)** | 75.21 **(0)**
 :green_circle: | app/client/src/widgets/CanvasWidget.tsx | 93.33 **(15.55)** | 66.67 **(13.34)** | 87.5 **(12.5)** | 95 **(12.5)**
 :green_circle: | app/client/src/widgets/ContainerWidget.tsx | 96.43 **(1.79)** | 77.78 **(5.56)** | 90 **(0)** | 96.23 **(1.89)**
 :green_circle: | app/client/src/widgets/FormButtonWidget.tsx | 67.5 **(17.5)** | 32 **(32)** | 55.56 **(22.23)** | 66.67 **(17.95)**
 :green_circle: | app/client/src/widgets/FormWidget.tsx | 86.27 **(50.98)** | 52.63 **(52.63)** | 84.62 **(69.24)** | 87.23 **(51.06)**
 :green_circle: | app/client/src/widgets/MetaHOC.tsx | 82.35 **(7.35)** | 57.89 **(10.52)** | 92.31 **(0)** | 83.05 **(8.47)**
 :green_circle: | app/client/src/widgets/TextWidget.tsx | 94.44 **(5.55)** | 50 **(50)** | 80 **(20)** | 94.12 **(5.88)**
 :green_circle: | app/client/src/widgets/ChartWidget/index.tsx | 92.59 **(18.52)** | 66.67 **(66.67)** | 77.78 **(44.45)** | 92 **(16)**
 :green_circle: | app/client/src/widgets/ListWidget/ListComponent.tsx | 93.33 **(26.66)** | 66.67 **(5.56)** | 83.33 **(50)** | 93.33 **(26.66)**
 :green_circle: | app/client/src/widgets/ListWidget/ListWidget.tsx | 52.24 **(16.79)** | 33.12 **(12.1)** | 43.48 **(15.22)** | 50.79 **(16.14)**
 :green_circle: | app/client/test/factories/Widgets/ListFactory.ts | 100 **(25)** | 100 **(0)** | 100 **(100)** | 100 **(25)**</details>

### CODE REVIEW

File: app/client/src/reducers/uiReducers/canvasSelectionReducer.ts [hunk 1]
```diff
@@ -1,8 +1,12 @@
 import { createImmerReducer } from "utils/AppsmithUtils";
 import { ReduxAction, ReduxActionTypes } from "constants/ReduxActionConstants";
+import { MAIN_CONTAINER_WIDGET_ID } from "constants/WidgetConstants";
+import { XYCoord } from "react-dnd";
```
Comment (riodeuno) [inline app/client/src/reducers/uiReducers/canvasSelectionReducer.ts:4]: I was meaning to ask this. Are we using react-dnd anymore? If we aren't can we remove this library from our codebase and use the `XYCoord` defined by us?

File: app/client/src/pages/common/CanvasSelectionArena.tsx [hunk 2]
```diff
@@ -306,21 +385,60 @@ export function CanvasSelectionArena({
...
+            canvasRef.current.height =
+              (snapRows * snapRowSpace +
+                (widgetId === MAIN_CONTAINER_WIDGET_ID
+                  ? theme.canvasBottomPadding
+                  : 0)) *
+              scale;
+          }
+          canvasCtx = canvasRef.current.getContext("2d");
+          canvasCtx.scale(scale, scale);
+          removeEventListeners();
```
Comment (riodeuno) [inline app/client/src/pages/common/CanvasSelectionArena.tsx:420]: Is this necessary? As we're running `removeEventListeners` in the cleanup phase.

**Review by riodeuno — DISMISSED** [review]
This needs QA

File: app/client/src/pages/common/CanvasSelectionArena.tsx (hunk 2 above)
Comment (marks0351) [reply app/client/src/pages/common/CanvasSelectionArena.tsx:420]: not really i think. but just a fail safe if there was any error during unmount.

**Comment by YogeshJayaseelan** [comment]
@marks0351 Verified #6102 PR and now user is able to select all sub widgets inside container/Form/Tabs using cmd+a

**Comment by YogeshJayaseelan** [comment]
@marks0351 Verified the fix for #5995 and now user is able to drag from outside editor and select widgets

**Review by riodeuno — APPROVED** [review]

**Pull request merged**